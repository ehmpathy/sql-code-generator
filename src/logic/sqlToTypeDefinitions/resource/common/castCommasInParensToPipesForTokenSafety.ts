export const castCommasInParensToPipesForTokenSafety = ({ sql }: { sql: string }) => {
  // track state of whether we are in parenthesizes or not
  let areInsideParensDepth = 0; // init as 0, since not inside parens
  const areInsideParens = () => areInsideParensDepth > 0; // whenever areInsideParensDepth > 0, we are inside parens

  // init object to build up the "casted" string
  let castedString = '';

  // loop through each char and update state / replace char as needed
  [...sql].forEach((char) => {
    // 1. manage "areInsideParens" state
    if (char === '(') {
      areInsideParensDepth += 1; // one parens deeper!
    }
    if (char === ')') {
      if (!areInsideParens) throw new Error('exiting parenthesizes without having opened them; not supported'); // fail fast for unsupported case
      areInsideParensDepth -= 1; // mark that we've dropped off one parens now
    }

    // 2. cast the char, considering whether we're in parens or not
    const castedChar = areInsideParens() && char === ',' ? '|' : char; // cast "," if we are inside parens

    // 3. add casted char to string
    castedString += castedChar;
  });

  // return the casted string
  return castedString;
};
