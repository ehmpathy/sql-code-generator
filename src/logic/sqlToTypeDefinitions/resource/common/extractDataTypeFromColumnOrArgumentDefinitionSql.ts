import { DataType } from '../../../../model';

// https://dev.mysql.com/doc/refman/8.0/en/string-types.html
const mysqlStringTypes = [
  'CHAR',
  'VARCHAR',
  'TINYTEXT',
  'TEXT',
  'MEDIUMTEXT',
  'LONGTEXT',
  'ENUM', // note: we treat enums as strings; TODO: https://github.com/uladkasach/sql-code-generator/issues/1
];

// https://www.postgresql.org/docs/9.5/datatype-character.html
const postgresStringTypes = [
  'CHARACTER',
  'CHAR',
  'CHARACTER VARYING',
  'VARCHAR',
  'TEXT',
  'UUID', // this too: https://www.postgresql.org/docs/9.5/datatype-uuid.html
];

// combine the types so we can search over the union
const dbStringTypes = [...mysqlStringTypes, ...postgresStringTypes];

// https://dev.mysql.com/doc/refman/8.0/en/numeric-types.html
const mysqlNumberTypes = [
  'INTEGER',
  'INT',
  'SMALLINT',
  'TINYINT',
  'MEDIUMINT',
  'BIGINT',
  'DECIMAL',
  'NUMERIC',
  'FLOAT',
  'DOUBLE',
];

// https://www.postgresql.org/docs/9.5/datatype-numeric.html
const postgresNumberTypes = [
  'SMALLINT',
  'INT2',
  'INTEGER',
  'INT',
  'INT4',
  'BIGINT',
  'INT8',
  'DECIMAL',
  'NUMERIC',
  'REAL',
  'DOUBLE PRECISION',
  'FLOAT8',
  'SMALLSERIAL',
  'SERIAL2',
  'SERIAL',
  'SERIAL4',
  'BIGSERIAL',
  'SERIAL8',
];

// combine the types so we can search over the union
const dbNumberTypes = [...mysqlNumberTypes, ...postgresNumberTypes];

// https://dev.mysql.com/doc/refman/8.0/en/date-and-time-types.html
const mysqlDateTypes = ['DATE', 'TIME', 'DATETIME', 'TIMESTAMP', 'YEAR'];

// https://www.postgresql.org/docs/9.5/datatype-datetime.html
const postgresDateTypes = [
  'TIMESTAMP',
  'TIMESTAMPTZ',
  'TIMESTAMP WITH TIMEZONE',
  'DATE',
  'TIME',
  'TIMETZ',
  'TIME WITH TIMEZONE',
];

// combine the types so we can search over the union
const dbDateTypes = [...mysqlDateTypes, ...postgresDateTypes];

// https://dev.mysql.com/doc/refman/8.0/en/binary-varbinary.html
const mysqlBinaryTypes = ['BINARY', 'VARBINARY'];

// https://www.postgresql.org/docs/9.5/datatype-binary.html
const postgresBinaryTypes = ['BYTEA'];

// combine the types so we can search over the union
const dbBinaryTypes = [...mysqlBinaryTypes, ...postgresBinaryTypes];

export const extractDataTypeFromColumnOrArgumentDefinitionSql = ({ sql }: { sql: string }) => {
  // 1. cast to upper
  const sqlUpper = sql.toUpperCase();

  // 2. one by one try to see which datatype is being defined in this definition
  if (dbStringTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.STRING;
  if (dbNumberTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.NUMBER;
  if (dbDateTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.DATE;
  if (dbBinaryTypes.some((dbType) => sqlUpper.includes(` ${dbType}`))) return DataType.BUFFER;
  throw new Error(`could not extract data type from '${sql}'`);
};
