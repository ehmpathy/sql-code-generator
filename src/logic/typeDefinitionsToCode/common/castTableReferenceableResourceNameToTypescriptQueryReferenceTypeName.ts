import { ResourceType } from '../../../model';
import { castResourceNameToTypescriptTypeName } from './castResourceNameToTypescriptTypeName';

export const castTableReferenceableResourceNameToTypescriptQueryReferenceTypeName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  // 0. throw error if not view or table, only view and table are possible table references
  if (![ResourceType.TABLE, ResourceType.VIEW].includes(resourceType)) {
    throw new Error('not a table referenceable query');
  }

  // 1. get the regular typescript type name
  const typescriptTypeName = castResourceNameToTypescriptTypeName({ name, resourceType });

  // 2. return the referenceable name
  if (resourceType === ResourceType.TABLE) return typescriptTypeName.replace(/^SqlTable/, 'SqlTableOrView');
  if (resourceType === ResourceType.VIEW) return typescriptTypeName.replace(/^SqlView/, 'SqlTableOrView');
  throw new Error('unexpected'); // this should never occur, but fail fast
};
