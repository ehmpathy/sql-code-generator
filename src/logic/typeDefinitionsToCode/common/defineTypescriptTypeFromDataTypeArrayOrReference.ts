import {
  DataType,
  TypeDefinitionReference,
  TypeDefinition,
  TypeDefinitionOfQueryTableReference,
} from '../../../domain';
import { defineTypescriptTypeFromDataTypeArray } from './defineTypescriptTypeFromDataTypeArray';
import { defineTypescriptTypeFromReference } from './defineTypescriptTypeFromReference/defineTypescriptTypeFromReference';

export const defineTypescriptTypeFromDataTypeArrayOrReference = ({
  type,
  plural,
  queryTableReferences,
  typeDefinitions,
}: {
  type: DataType[] | TypeDefinitionReference;
  plural: boolean;
  typeDefinitions: TypeDefinition[];
  queryTableReferences: TypeDefinitionOfQueryTableReference[];
}) => {
  // if its a type reference, then use that handler
  if (type instanceof TypeDefinitionReference) {
    const defined = defineTypescriptTypeFromReference({
      reference: type,
      queryTableReferences,
      typeDefinitions,
    });
    if (plural) return `${defined}[]`;
    return defined;
  }

  // else, it must be data array. use that handler
  return defineTypescriptTypeFromDataTypeArray({ type });
};
