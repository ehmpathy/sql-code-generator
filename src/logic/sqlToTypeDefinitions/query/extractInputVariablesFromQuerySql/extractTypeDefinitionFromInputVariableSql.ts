import { DataType } from '../../../../domain';
import { TypeDefinitionOfQueryInputVariable } from '../../../../domain/objects/TypeDefinitionOfQueryInputVariable';
import { TypeDefinitionReference } from '../../../../domain/objects/TypeDefinitionReference';
import { throwErrorIfTableReferencePathImpliesTable } from '../common/throwErrorIfTableReferencePathImpliesTable';

export const extractTypeDefinitionFromInputVariableSql = ({
  token,
  sql,
}: {
  token: string;
  sql: string;
}): TypeDefinitionOfQueryInputVariable => {
  // define regex for token w/ optional type casting
  const tokenWithOptionalTypecastingRegexString = `${token}(?:\\:\\:\\w+(?:\\[\\])?)?`;

  // check if this token matches the "resource.column =,<=,>= :token" pattern; if so, then the input type = the resource.column type
  const [
    _, // tslint:disable-line no-unused
    leftEqualsTableReferencePath, // check if "resource.column = :token" or "resource.column <= :token" or "resource.column <= :token"
  ] =
    new RegExp(
      `(\\w+\\.?\\w*)\\s?(?:<|>)?=\\s?(?:${tokenWithOptionalTypecastingRegexString})(?:[^\\w]|$)`,
      'g',
    ).exec(sql) ?? [];
  if (leftEqualsTableReferencePath) {
    throwErrorIfTableReferencePathImpliesTable({
      referencePath: leftEqualsTableReferencePath,
    });
    return new TypeDefinitionOfQueryInputVariable({
      name: token.replace(':', ''),
      type: new TypeDefinitionReference({
        tableReferencePath: leftEqualsTableReferencePath,
        functionReferencePath: null,
      }),
      plural: false,
    });
  }

  // check if this token matches the ":token =,<=,>= resource.column" pattern; same as above, but flipped
  const [
    __, // tslint:disable-line no-unused
    rightEqualsTableReferencePath, // check if ":token = resource.column" or ":token <= resource.column" or ":token >= resource.column"
  ] =
    new RegExp(
      `(?:${tokenWithOptionalTypecastingRegexString})\\s?(?:<|>)?=\\s?(\\w+\\.?\\w*)`,
    ).exec(sql) ?? [];
  if (rightEqualsTableReferencePath) {
    throwErrorIfTableReferencePathImpliesTable({
      referencePath: rightEqualsTableReferencePath,
    });
    return new TypeDefinitionOfQueryInputVariable({
      name: token.replace(':', ''),
      type: new TypeDefinitionReference({
        tableReferencePath: rightEqualsTableReferencePath,
        functionReferencePath: null,
      }),
      plural: false,
    });
  }

  // check if this token matches the "resource.column ANY (:token)" pattern; if so, then the input type = an array of the resource.column type
  const [
    ___, // tslint:disable-line no-unused
    leftEqualsArrayTableReferencePath, // check if "resource.column = :token" or "resource.column <= :token" or "resource.column <= :token"
  ] =
    new RegExp(
      `(\\w+\\.?\\w*)\\s?=?\\s?(?:any|ANY|in|IN)\\s?\\(\\s?(?:${tokenWithOptionalTypecastingRegexString})\\s?\\)(?:[^\\w]|$)`,
      'g',
    ).exec(sql) ?? [];
  if (leftEqualsArrayTableReferencePath) {
    throwErrorIfTableReferencePathImpliesTable({
      referencePath: leftEqualsArrayTableReferencePath,
    });
    return new TypeDefinitionOfQueryInputVariable({
      name: token.replace(':', ''),
      type: new TypeDefinitionReference({
        tableReferencePath: leftEqualsArrayTableReferencePath,
        functionReferencePath: null,
      }),
      plural: true,
    });
  }

  // check if this token is used in a function. If so, its equivalent to whatever is at that index of the function params
  const reg = `\\s+(\\w+\\((?:\\s*[:\\w#]+,)*(?:\\s*${tokenWithOptionalTypecastingRegexString},?)(?:\\s*[:\\w#]+,?)*\\s?\\))`; // note: this reg matches the whole function def (e.g., `upsert_image(:url,:caption,:credit)`)
  const [
    // eslint-disable-next-line @typescript-eslint/naming-convention, @typescript-eslint/no-unused-vars
    ____,
    tokenInsideFunctionMatch, // check if "functionName(arg1?, :token, arg2?)"
  ] = new RegExp(reg).exec(sql) ?? [];
  if (tokenInsideFunctionMatch) {
    // grab the function name
    const functionName = tokenInsideFunctionMatch.split('(')[0]!.trim();

    // figure out where in the array the token is
    const sqlAfterOpen = tokenInsideFunctionMatch.split('(')[1]!;
    const sqlBetweenOpenAndClose = sqlAfterOpen.split(')')[0]!;
    const parametersArray = sqlBetweenOpenAndClose
      .split(',')
      .map((str) => str.trim());
    const index = parametersArray.findIndex((parameter) =>
      new RegExp(tokenWithOptionalTypecastingRegexString).test(parameter),
    );
    if (index === -1) throw new Error('unexpected'); // this should never occur if the regex is working correctly; fail fast though

    // return the function path
    return new TypeDefinitionOfQueryInputVariable({
      name: token.replace(':', ''),
      type: new TypeDefinitionReference({
        functionReferencePath: `${functionName}.input.${index}`,
        tableReferencePath: null,
      }),
      plural: false,
    });
  }

  // check if this token is used in a limit. If so, type = a number
  const tokenUsedForLimit = new RegExp(
    `(?:LIMIT|limit)\\s+${token}(?:[^\\w]|$)`,
  ).test(sql); // check if "LIMIT :token"
  if (tokenUsedForLimit) {
    return new TypeDefinitionOfQueryInputVariable({
      name: token.replace(':', ''),
      type: [DataType.NUMBER],
      plural: false,
    });
  }

  // check if this token is used in an offset. If so, type = a number
  const tokenUsedForOffset = new RegExp(
    `(?:OFFSET|offset)\\s+${token}(?:[^\\w]|$)`,
  ).test(sql); // check if "OFFSET :token"
  if (tokenUsedForOffset) {
    return new TypeDefinitionOfQueryInputVariable({
      name: token.replace(':', ''),
      type: [DataType.NUMBER],
      plural: false,
    });
  }

  // check if this token is used in a null comparison. if so, type = null
  const tokenUsedForNullComparison = new RegExp(
    `(?:${tokenWithOptionalTypecastingRegexString})\\s+(?:IS|is)\\s+(?:NULL|null)(?:[^\\w]|$)`,
  ).test(sql); // check if "OFFSET :token"
  if (tokenUsedForNullComparison) {
    return new TypeDefinitionOfQueryInputVariable({
      name: token.replace(':', ''),
      type: [DataType.NULL],
      plural: false,
    });
  }

  // if none of the above worked, throw an error. one should work...
  throw new Error(
    `could not extract type definition for input variable '${token}'`,
  );
};
