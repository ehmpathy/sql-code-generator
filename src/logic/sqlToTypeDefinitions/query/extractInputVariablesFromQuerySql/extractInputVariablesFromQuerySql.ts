import strip from 'sql-strip-comments';

import { flattenSqlByReferencingAndTokenizingSubqueries } from '../common/flattenSqlByReferencingAndTokenizingSubqueries';
import { extractInputVariableTokensFromQuerySql } from './extractInputVariableTokensFromQuerySql';
import { extractTypeDefinitionFromInputVariableSql } from './extractTypeDefinitionFromInputVariableSql';

/**
 * goal:
 *  - try to define input type when possible
 * - fallback to "any" when not possible and ask user to add an issue w/ use case when this occurs
 *   - so that the generator is non-blocking
 *
 *  e.g.,:
 *   - where u.id = :id => "{ id: number }"
 *   - where lower(u.name) = :name => { name: any } (for mvp);
 *     - later, { name: string } (since we'd track standard fn's; TODO)
 *
 * steps:
 * - 1. find all of the `:__variable_name__` pattern input definitions (yesql format)
 * - 2. attempt to figure out the type for each variable name
 */

export const extractInputVariablesFromQuerySql = ({ sql }: { sql: string }) => {
  // find each input variable that needs to have a type def defined
  const inputVariableTokens = extractInputVariableTokensFromQuerySql({ sql });

  // flatten the sql, to avoid subquery syntax interfering with below search patterns
  const strippedSql = strip(sql);
  const { flattenedSql, references: subqueries } = flattenSqlByReferencingAndTokenizingSubqueries({
    sql: strippedSql,
  });

  // define the type definition for each variable
  const definitions = inputVariableTokens.map((token) => {
    // check if the token is actually in a subquery; run it on the subquery if so
    const foundSubqueryContainingToken = subqueries.find((subquery) => subquery.sql.includes(token));
    if (foundSubqueryContainingToken)
      return extractTypeDefinitionFromInputVariableSql({ token, sql: foundSubqueryContainingToken.sql });

    // otherwise, run it on the root flattened query (flattened to make sure subquery syntax does not cause problems)
    return extractTypeDefinitionFromInputVariableSql({ token, sql: flattenedSql });
  });

  // return defs
  return definitions;
};
