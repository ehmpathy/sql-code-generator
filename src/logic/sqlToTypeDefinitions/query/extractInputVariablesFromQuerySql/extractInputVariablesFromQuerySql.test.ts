import { extractSqlFromFile } from '../../../common/extractSqlFromFile';
import { extractInputVariablesFromQuerySql } from './extractInputVariablesFromQuerySql';

describe('extractSelectExpressionsFromQuerySql', () => {
  it('should be able to determine types accurately for this example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/find_image_by_id.sql`,
    });
    const defs = extractInputVariablesFromQuerySql({ sql });
    expect(defs.length).toEqual(1);
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this other example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/select_suggestion.sql`,
    });
    const defs = extractInputVariablesFromQuerySql({ sql });
    expect(defs.length).toEqual(0);
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for this other other example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/upsert_suggestion.sql`,
    });
    const defs = extractInputVariablesFromQuerySql({ sql });
    expect(defs.length).toEqual(5);
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for yet another example', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/find_all_suggestions_by_normalized_source.sql`,
    });
    const defs = extractInputVariablesFromQuerySql({ sql });
    expect(defs.length).toEqual(1);
    expect(defs).toMatchSnapshot();
  });
  it('should be able to determine types accurately for a case where an input variable is defined in a subquery used as an input to a function', async () => {
    const sql = await extractSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/queries/upsert_profile_with_subselect.sql`,
    });
    const defs = extractInputVariablesFromQuerySql({ sql });

    expect(defs.length).toEqual(4);
    expect(defs).toMatchSnapshot();
  });
});
