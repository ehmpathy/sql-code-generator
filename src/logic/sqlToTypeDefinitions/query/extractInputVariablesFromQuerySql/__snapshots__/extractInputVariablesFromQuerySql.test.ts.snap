// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for a case where an input variable is defined in a subquery used as an input to a function 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "serviceProviderUuid",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "p.uuid",
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "bannerImageUuid",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_provider_profile.input.1",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "pictureImageUuid",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_provider_profile.input.2",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "introduction",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_provider_profile.input.3",
      "tableReferencePath": null,
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for this example 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "id",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "i.id",
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for this other example 1`] = `[]`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for this other other example 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "suggestionSource",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.input.0",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "externalId",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.input.1",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "suggestedIdeaId",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.input.2",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "status",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.input.3",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "result",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_suggestion.input.4",
      "tableReferencePath": null,
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to determine types accurately for yet another example 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "name",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "get_id_from_suggestion_source_name.input.0",
      "tableReferencePath": null,
    },
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to extract input variables from a function that compares a timestamp with less-than-or-equals-to and null 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "threadUuid",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "chat_thread.uuid",
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "until",
    "plural": false,
    "type": [
      "null",
    ],
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "until",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "chat_message.created_at",
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "limit",
    "plural": false,
    "type": [
      "number",
    ],
  },
]
`;

exports[`extractSelectExpressionsFromQuerySql should be able to extract input variables from a function that postgres::typecasts some of the input variables 1`] = `
[
  TypeDefinitionOfQueryInputVariable {
    "name": "homeStationGeocodeId",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_train.input.0",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "combinationId",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_train.input.1",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "locomotiveUuids",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "unnest.input.0",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "carriageUuids",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "unnest.input.0",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "engineerUuids",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "unnest.input.0",
      "tableReferencePath": null,
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "leadEngineerUuid",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": null,
      "tableReferencePath": "train_engineer.uuid",
    },
  },
  TypeDefinitionOfQueryInputVariable {
    "name": "status",
    "plural": false,
    "type": TypeDefinitionReference {
      "functionReferencePath": "upsert_train.input.6",
      "tableReferencePath": null,
    },
  },
]
`;
